src/scheduler/node_info.c

```diff
@@ -637,13 +776,80 @@ void
 free_nodes(node_info **ninfo_arr)
 {
 	int i;
+	int chunk_size;
+	th_data_free_ninfo *tdata = NULL;
+	th_task_info *task = NULL;
+	int num_tasks = 0;
+	int num_nodes;
+	int thid;
 
-	if (ninfo_arr != NULL) {
-		for (i = 0; ninfo_arr[i] != NULL; i++)
-			free_node_info(ninfo_arr[i]);
+	if (ninfo_arr == NULL)
+		return;
 
+	num_nodes = count_array((void **) ninfo_arr);
+
+	thid = *((int *) pthread_getspecific(th_id_key));
+	if (thid != 0) { /* worker thread, don't use multithreading */
+		tdata = malloc(sizeof(th_data_free_ninfo));
+		if (tdata == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			return;
+		}
+		tdata->ninfo_arr = ninfo_arr;
+		tdata->sidx = 0;
+		tdata->eidx = num_nodes - 1;
+		free_node_info_chunk(tdata);
+		free(tdata);
 		free(ninfo_arr);
+		return;
 	}
+	i = 0;
+	chunk_size = num_nodes / num_threads;
+	chunk_size = (chunk_size > 1024) ? chunk_size : 1024;
+	while (num_nodes > 0) {
+		tdata = malloc(sizeof(th_data_free_ninfo));
+		if (tdata == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			break;
+		}
+
+		tdata->ninfo_arr = ninfo_arr;
+		tdata->sidx = i;
+		i += chunk_size;
+		tdata->eidx = i - 1;
+		task = malloc(sizeof(th_task_info));
+		if (task == NULL) {
+			free(tdata);
+			log_err(errno, __func__, MEM_ERR_MSG);
+			break;
+		}
+		task->task_type = TS_FREE_ND_INFO;
+		task->thread_data = (void *) tdata;
+
+		pthread_mutex_lock(&work_lock);
+		ds_enqueue(work_queue, (void *) task);
+		pthread_cond_signal(&work_cond);
+		pthread_mutex_unlock(&work_lock);
+
+		num_nodes -= chunk_size;
+		num_tasks++;
+	}
+
+	/* Get results from worker threads */
+	for (i = 0; i < num_tasks;) {
+		pthread_mutex_lock(&result_lock);
+		while (ds_queue_is_empty(result_queue))
+			pthread_cond_wait(&result_cond, &result_lock);
+		pthread_mutex_unlock(&result_lock);
```

Comment:
the unlock does not sound correct here. cond_wait will re-acquire locks atomically at return. You are not using that before checking ds_queue_is_empty in the first while loop - just doing it after the lock...besides the lower while loop seems unprotected in its use of result_queue

Reply:
Thanks for catching this Subhasis!